Refactor to reduce cyclomatic complexity, add tests

grimxreaper opened the pull request:
This branch introduces a refactoring of app.js in order to reduce cyclomatic complexity which was making the code difficult to reason about. It also improves the test's names so that what each test is testing is more clear.

- Removes duplicate logic and code
- Inlines all functions inside of onClick to reduce the number of places where result is being set inside of setState
- Fixes new bugs that were not detected before and created new tests

bryzaguy commented:
Thanks for getting this created! Cleaning up this PR will help me discover suggestions.

Please look through the diff and remove any commented tests, console logs, added spaces or removed lines.

Why is the switch js file included in the commit when it isn't used? Can we remove it?

A PR is best thought of as a proposal. It's your opportunity to convince your team that the code needs to change. Your PR describes what code changes it contains but not why it's a good idea. Typically, the less changes a PR has, the better. So when it's a big change, details of what changed are usually in support of why they are needed.

The rule for commit messages is to complete the sentence "Applying this commit will _______". Only the part you fill in should be in the message. "This commit..." should not be in the commit message. The idea is that the message reads in present tense saying what it does. The very first commit follows this rule. All the others are inconsistent. See if you can use git rebase to update your commit messages.
<img width="629" alt="image" src="https://user-images.githubusercontent.com/2603435/119790791-0f372000-be89-11eb-8dea-8060f443089d.png">

bryzaguy commented:
Some of your test names are really good! Some of them need work. Updating this will help me understand your coverage better. Good test names have a "Given, When, Then" format like "When <given context>, <when action happens> <then result occurs>". In your tests, you wrap some in a "describe" and present the Given part (e.g. "When result is empty..."), which is a nice optimization but your nested tests should complete the sentence. For example, "division version" could be "for division" making it read as "When result is empty should default to zero for division". Does that makes sense? 

I should be able to read your tests and know what they are testing, without being burdened by their internals (e.g. test names with "7 * 2" vs "multiply two numbers"). "Must return 6 parens and return 874" is all internal detail! What is the behavior of the calculator that you want other developers to know should keep working? Rename the test with the answer to that question.

Some of your test names say "unless"... are they testing two things? Split those tests up if so, or take out what they aren't testing.

What is "should equal an empty string" testing?

<img width="870" alt="image" src="https://user-images.githubusercontent.com/2603435/119795064-f597d780-be8c-11eb-98f2-46f6317bf768.png">

grimxreaper commented:
@bryzaguy thank you!

I completed all suggested changes with app.js and with the tests.

Regarding the commit messages, first I did a test with git rebase interactive and changed the first few commit messages successfully. Then, I went through all 22+ commit messages and changed each one but when wanting to save, git told me there were conflicts, and even when I would fix them, on the next one it would ask me to fix the same conflicts I had already fixed and staged, and change the commit message over and over again, but would not save, I became very nervous about accepting current changes because I don't want to incorrectly alter the git commits themselves just the messages, and so I aborted after re-trying several times and not understanding the mistake.

bryzaguy commented:
Why is your last commit a merge? Did you make changes through the GitHub site? Reset the last commit and commit it again then force push (don't pull first)

The test names are improving. Not seeing changes for some of my suggestions/questions. Double check that the test names clearly convey what they're testing.

There should be no merges to deal with in your rebase. Did you you follow steps like this? https://linuxize.com/post/change-git-commit-message/

Still seeing the switch JS file.

Please update the title and description.

grimxreaper commented:
@bryzaguy I did not do it on github, I did it in the terminal, yes I followed those steps exactly. I have no idea why this happened because I was extremely careful and working slowly to follow each step to the T. I tested whether the git rebase interactive would work by changing the first few commit messages, that worked. After that, once I did the rest, (changed all the commit messages that you do not see reflected here because they never were accepted and pushed) it did not let me force push, and it asked me to fix conflicts, in order to be able to push, it also asked me to pull changes because after changing all the commit messages it told me I am somehow behind on 21 commits, and ahead on 21 commits at the same time, so that is why I did that merge because it instructed me to as a solution. Then, like I said, it brought up merge conflicts for every single name change after that, and because I did know which changes to accept for each, and was nervous about actually changing the git history rather than just the commit name I stopped. At present, it is not letting me do a git reset because I had closed out of that terminal hours ago...not sure what to do, do you recommend doing a git rebase to remove that merge? Hmm...I changed the majority of the test names...are those not depicting what the test is testing?

bryzaguy commented:
It said you were 21 commits behind because git rebase changes your history to a different history. Once you force push the remote and local repo will have matching commits. Is it possible you had uncommitted changes before starting rebase? Rebase takes you back in time so you want to make sure everything is in a clean state.

bryzaguy commented:
The title and description have improved. The title is missing why the refactor is important. The description should tell me why you inlined into onClick if you mention it. You don't need to mention fixing things you broke during refactoring.

bryzaguy commented:
What is "should equal an empty string" testing?

Is "a subtraction followed by an addition defaults to an addition and returns correct result" two tests that need to be split? Is the latter covered elsewhere?

Is "a new calculation begins if keys are pressed unless equals is pressed again for whole numbers" two tests? Reads like two different cases.

What is the given, when, then of "multiplies the result of operations done inside parentheses"?

Does "check the operation of 2 numbers division of two numbers returns correct result" read like a sentence?

I'm really hoping you'll organize related tests into describe blocks that setup your given context and make sure test titles are easy to understand because they read like sentences. This will help when trying to spot redundant tests and gaps in coverage.

bryzaguy commented on src/App.js:
```diff
@@ -7,350 +7,148 @@ import { evaluate, round } from "mathjs";
+    const isEmptyLastOper =
+      (button.key === "*" || button.key === "/") && this.state.result === "";
+
+    const replaceLastCharWithCurrentKey =
```
Can we extract "replaceLastCharWithCurrentKey" as a function to reduce the size of our onClick?

bryzaguy replied:
@grimxreaper ^

grimxreaper replied:
So I extracted it, and in the console it shows the correct result, but it does not reflect it on the display, have been trying to figure out why for an hour, but not sure why it is doing that.

bryzaguy replied:
Are you passing the button key to your new function and running it every click?

grimxreaper replied:
Yes I tried that as well, it's interesting because even before changing anything, in the console it displayed the same functionality as after I changed it to the new function, but for some reason it doesn't reflect it on the display with the new function, and it doesn't seem to have anything to do with state.

bryzaguy replied:
Not sure I understand but I know you got this!

grimxreaper replied:
So when tempResult finishes being sliced and comes back in this if block, it is reset to what it was as though slice didn't happen. You're right, this shouldn't be hard but I can't seem to figure out why, I'll keep at it.

grimxreaper replied:
I haven't been able to figure this one out, I'm gonna focus on something else since I've devoted a lot of hours to it.

bryzaguy replied:
Have you tried adding console logs? Maybe try ```debugger;``` to walk through the code as you use the app in the browser? Most times we get stuck because we have some preconceived idea of where the problem is or isn't and we need to step back and just observe. I know you can do it.

bryzaguy replied:
Also, do your tests still pass? I would guess that they don't. Try starting over and pause after each tiny code change to catch the moment they fail. Spend time understanding why.

grimxreaper replied:
Finished with this change and reviewed the last commit to remove empty lines, console logs and making the tests pass, please let me know if it is satisfactory! :D @bryzaguy

bryzaguy commented on src/App.js:
```diff
@@ -7,350 +7,148 @@ import { evaluate, round } from "mathjs";
+    const isMinusLastMinus = button.key === "-" && hasLastChar("-");
+    const isMinusLastAdd = button.key === "+" && hasLastChar("-");
+    const isEmptyLastOper =
+      (button.key === "*" || button.key === "/") && this.state.result === "";
```
Why wouldn't you want to default to zero for plus or minus?

grimxreaper replied:
It already does that by default.

bryzaguy commented on src/App.js:
```diff
@@ -7,350 +7,148 @@ import { evaluate, round } from "mathjs";
+        const results = lastFormula.match(operatorAndNumRegex) || [];
+        const lastNum = results[results.length - 1];
+        const operate = lastNum[0]; // "*3.3" => "*"
+        const originalLastNum = lastNum.substr(1); // "*3.3" => "3.3"
```
Can we extract a function that returns operate and originalLastNum? Could those use better names?

bryzaguy replied:
@grimxreaper ^

grimxreaper replied:
I've been attempting to but how would I pull in `lastFormula` into the new file since I need it for results?

bryzaguy replied:
Couldn't you pass it as an argument?

grimxreaper replied:
I did but it comes up as undefined.
`[TypeError: Cannot read property 'match' of undefined]`

grimxreaper replied:
The bug with this one looks very similar to the one for `replaceLastCharWithCurrentKey`, they both do not keep the variable updated when re-entering App.js after completing their functions in their own files. I'm not sure how to fix this bug given that it seems to be the same one I haven't been able to figure out.

bryzaguy commented on src/App.js:
```diff
@@ -7,350 +7,148 @@ import { evaluate, round } from "mathjs";
+      const pAndDigit = /(\))([0123456789])/g;
+      const pAndp = /(\))(\()/g;
 
+      newTempResult = newTempResult
```
Can we extract these mutations to tempResult with functions named so it's clear what they are trying to achieve / why they're necessary? I suggest something like
```
tempResult = mutuateForReasons(tempResult)
```
So that you still see tempResult is replaced without having to go into the function.

bryzaguy replied:
@grimxreaper ^

bryzaguy commented:
Okay I've added some code comments. If you keep running into issues with rebase let's punt on it. I'll be really impressed if you figure it out though. Please let me know when you've been able to finish the changes I've suggested. Good luck!

grimxreaper commented:
- [x]  What is "should equal an empty string" testing?

I don't think that test name was clear so I changed it, it was testing the display, and I added it inside a describe block with other related tests that are only testing what result is being displayed.

- [x] Is "a subtraction followed by an addition defaults to an addition and returns correct result" two tests that need to be split? Is the latter covered elsewhere?

I wouldn't think that test needs to be split because it is testing what I want it to test and it includes the result.

- [x] Is "a new calculation begins if keys are pressed unless equals is pressed again for whole numbers" two tests? Reads like two different cases.

I agree, I added another describe block in it and added another test that didn't have a describe block within this outer describe block, curious to know if you think this approach reads better.

- [x] What is the given, when, then of "multiplies the result of operations done inside parentheses"?

- I added that in a describe block as 'when operations are in parentheses'... "and follow each other, then their results are multiplied together".
- I also added describe blocks for all the tests and changed the names so it would read as you mentioned with the given, when, then. I like the names much better now I think they read better like sentences.

- [x] Does "check the operation of 2 numbers division of two numbers returns correct result" read like a sentence?

I changed this as well, it reads at least minimally better as before it did not at all.

*I also removed one of the tests that was redundantly testing the same thing as another one.

*Please let me know if I made mistakes in my naming, like if I am not supposed to use the word 'given' like 'given the operation of 2 numbers...'.  Or if my naming is too lengthy. Thank you, I'm learning so much.

*I tried a git reset, git revert, and nothing seemed to work in returning to the commit before the merge, at one point it did seem to work, but then when I tried to reword the commit messages, it still gave me conflicts to resolve every time so I aborted.

bryzaguy commented:
By testing the result of "subtraction followed by an addition defaults to an addition and returns correct result" you're testing that addition defaults correctly AND that addition works. If you've already covered addition working, you'll have multiple failing tests if it breaks. Is there a way you can check that addition was defaulted without testing the result?

bryzaguy commented:
Your title and description are almost perfect. Why the "extract variables" bullet?

bryzaguy commented:
Okay thanks for trying to rename the commits. We will just have to try and catch it earlier next time. Or hopefully you'll remember the rule. :)

bryzaguy commented:
An alternative title could be "Refactor to reduce cyclomatic complexity, add tests"

bryzaguy commented:
Can you see my code comments?

grimxreaper commented:
- [x] Is there a way you can check that addition was defaulted without testing the result?
Yes, I changed it to only test that, I hadn't thought of that, thank you!

- [x] Why the "extract variables" bullet?
I was wondering whether to keep that, I guess it really isn't necessary as refactoring implies such changes.

- [x] Or hopefully you'll remember the rule. :)
Yes! I most definitely will because I got the practice of renaming all of them even if it didn't save them.

- [x] Can you see my code comments?
Which ones? I don't see any in the code, I thought you meant the ones here in the conversation since I didn't see any.

bryzaguy commented:
When you extract functions from onClick, put them into their own files and import them. Try to search for opportunities to extract functions from onClick so it becomes super readable. Like in the equalsPressed, pulling out functions where you'd have to study it to know what it's doing like when running the calculation to generate the new result value. Avoid extracting ```setState``` so it can stay in the onClick.

bryzaguy commented:
The last commit message _almost_ follows the rule but has the wrong tense.

grimxreaper commented:
> The last commit message _almost_ follows the rule but has the wrong tense.

Ah, yes I see! I successfully rebased but when I tried force push to the remote, it says it cannot read and wants me to git pull even though I had just made sure to push any changes prior to rebasing.

<img width="528" alt="Screen Shot 2021-05-29 at 3 57 55 PM" src="https://user-images.githubusercontent.com/64915527/120083492-0a47bb80-c097-11eb-8f24-1b453c7f21c9.png">

bryzaguy commented:
Don't pull!!
You are missing the `--force` for `git push --force`

grimxreaper commented:
no I added the --force, that's one of the lines you see underlined in pink, as per the instructions here:  https://linuxize.com/post/change-git-commit-message/

Also, yes I am not pulling, but instead of following those instructions and including the branch name, I just did `git push --force`, and I see it updated the commit message successfully.

bryzaguy commented:
Awesome! You need to include origin (if that's what you have your remote labeled) when pushing to GitHub. For example `git push --force origin branch-name`

bryzaguy commented:
The `from` portion of your imports should be the file name and it is definitely case sensitive and always need to match. You had enough things to correct so I didn't say anything before but putting these functions into you Components folder doesn't really make sense. They aren't components so the PascalCase doesn't make sense either. Components should have React components and maybe functions that are used by the components. Either leave them in the `src` folder or organize them into folders that say how they relate, like `evaluate` could be a folder that has all functions related to pressing `=`. Also, be vigilant at keeping the name of the export, file, and import consistent. That's super important and will be appreciated by your team.

bryzaguy commented:
One thing to keep in mind is that longer commit messages are truncated in many places, rendering them less valuable. So it's a good idea to keep them as short as possible or have the most important info come first. For your last commit, the word `function` is unnecessary. It's easy to discover that in the diff. Additionally, since the function names are so long, there could have been a commit for each function where you only stage the relevant portions of App.js (using `git add --patch ./src/App.js`) for each commit. Alternatively, a single commit with the three functions which doesn't say each file name would work if you can guess what files are included by the commit message, such as `extract evaluation functions`. I realize having the word 'functions' in the previous commit message seems to contradict what I said before about removing 'function', but if you imagine yourself trying to find the commit that added one of those evaluation functions in a list of commits, hopefully you'll see how it's different and helpful in providing context.

grimxreaper commented:
re: commit messages

I found a way to apply your advice and combine your suggestions in a way that would satisfy what the commit message for that commit should reflect. I hope that's better, thank you and I'll keep that in mind from here on out!

bryzaguy commented:
How relevant is the word `two` in your commit message?

bryzaguy commented:
An alternative commit message is `extract pressedAnotherOperator, evaluation functions`. Notice I removed 'and' as well as moving the specific function name to the front so that worst case the word 'functions' gets truncated? Every character counts.

grimxreaper commented:
Ah okay, I just thought it would be too imprecise to not include the 'two' as I may end up extracting more functions from pressedEquals later on, and this gives the impression that I already extracted all of them. But I think I may be overthinking it! ha! I don't think your edit is misleading.

grimxreaper committed:
delete unused files

grimxreaper committed:
extract keypads

grimxreaper committed:
refactor button onClick, tests, extract functions

grimxreaper commented:
Hi @bryzaguy ! I completed the changes we spoke of, and checked through App.js again for possible refactoring opportunities but everything looks pretty good to me, please let me know what you think!

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,76 @@ import React from "react";
-  };
-
-
-  roundedResult = (expression) => {
-    const digitAfterComma = 13;
-    round(evaluate(expression), digitAfterComma);
-  };
 
-  addMultiplier = (expression) => {
-    //Find all pattern Digit + (
-    //expression ="1(2+6)+23(6+9)+(2+3)9"
-
-    const digitAndP = /([0123456789])(\()/g;
-    const pAndDigit = /(\))([0123456789])/g;
-    // )( -> )*(
-    const pAndp = /(\))(\()/g;
-
-    //return the value
-    return expression
-      .replace(digitAndP, "$1*$2")
-      .replace(pAndDigit, "$1*$2")
-      .replace(pAndp, "$1*$2");
-
-
-  };
-
-
-
-  calculate = () => {
-    const { result, operate, originalLastNum } = this.state;
-    let finalResult = 0;
-    var tempResult = result;
-
-    // Special case with expression ending with operator, after CE use
-    if ("-+*/".includes(this.getLastChar(result))) {
-      tempResult = result.slice(0, -1);
+    if (lastEquation !== "" && button.key !== "=") {
+      //start new equation
+      this.setState({ result: button.key });
     }
```
Is this code necessary? Is it handled by the last block with is displayable?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
It does seem necessary because taking it out then adds the new number that is entered to the end of the last result rather than starting a new equation. Are you suggesting I can add code to isDisplayable to replace this? Because to answer your question, it is not currently handled by it.

bryzaguy replied:
@grimxreaper What would we need to change in the isDisplayable section to remove it?

bryzaguy replied:
If you figure out what would need to happen, go forward with the change.

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,76 @@ import React from "react";
-      if (
-        button.key !== "CE" &&
-        button.key !== "=" &&
-        button.key !== "(" &&
-        button.key !== "AC" &&
-        button.key !== ")"
-      ) {
-        this.setState({
-          operate: button.key,
-          //result: this.state.result + button.key,
-          result: this.changeKeys(this.state.result, button.key),
-        });
-      }
-    }
-    if (button.key === "=") {
-      this.calculate();
-    } else if (button.key === "AC") {
-      this.reset();
-    } else if (button.key === "CE") {
-      this.backspace();
-    }
-  };
+    const { result, lastEquation } = this.state;
 
-  recordLastNum = (result) => {
-    const regex = /[-+/*]{0,}[0-9]{1,}[.]{0,1}[0-9]*/g;
-    let results = result.match(regex) || [];
-    if (results.length === 0) {
-      return "";
-    }
-    //We need to take the last one
-    let lastNum = results[results.length - 1];
-    // If there is no operator -> we return the value
-    if (!"-+/*".includes(lastNum[0])) {
-      return lastNum;
+    if (button.key !== "=") {
+      this.setState({
+        lastEquation: "",
+      });
     }
```
Is there a way we can remove this code?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
So I'm not sure but in experimenting with if it *is* possible to take it out, I ran into a bug that occurs when starting a new equation. I tested the different mutations of it with all the different operators and it occurs with all of them. The error message reads `TypeError: result.substr is not a function`. 

When examining the stack trace it seems as though `lastEquation` isn't actually being reset upon starting a new equation. To fix this I set the state of `lastEquation` to an empty string where the code for starting a new equation is, but it didn't fix this, and I think it's because it is the operator that needs to be targeted and reset. I believe this is why we were setting the state of `operator` in the last iteration of the code, and resetting it when a new equation began.

Not sure what to change to fix this, any thoughts?

<img width="478" alt="Screen Shot 2021-06-09 at 12 10 08 PM" src="https://user-images.githubusercontent.com/64915527/121390718-a4bbc080-c91b-11eb-834b-5787c9ce1857.png">

bryzaguy replied:
@grimxreaper So it looks like we can't just remove it without some code breaking. Do we know why the code breaks and what we would have to do to fix it? If you figure it out, go ahead and add the change.

grimxreaper replied:
@bryzaguy Actually this bug is present even without me removing it! I just happened to stumble upon it when I was experimenting with removing this code.

grimxreaper replied:
So I fixed this bug by updating result rather than `lastEquation`, because result is what `hasLastChar` is changing using `.substr` with.

bryzaguy replied:
@grimxreaper Great job tracking that down! I'm confused by your test added though. Was typing an equation then equals then starting another equation how you reproduced the bug? Based on your fix and the error, I would have guessed you only needed to type a number then an operator. Your commit message for a bug fix should lead with the fix, for example "fix result to number instead of string". Can we delete that block of code now? If not, what needs to change for that to happen? Once you figure it out add a commit!

grimxreaper replied:
Yes that's how the bug was reproduced, that is the bug in fact, we do need the equals in the test because that is part of what causes the bug which is reflected in the stack trace.

Okay great, I will update the commit message.

That block of code still serves the purpose of displaying/saving what buttons have been pressed within the new equation, removing it causes a bug where it doesn't show the whole equation but only the last pressed button. I will investigate removing it.

grimxreaper replied:
@bryzaguy okay it should be fixed now that I changed where `lastEquation` is updated!

bryzaguy commented on src/isDisplayable.js:
```diff
@@ -0,0 +1,10 @@
+  const isOperator = "/*-+".includes(button.key);
+  if (openP || closedP || numOrDecimal || isOperator) {
+    return true;
+  }
```
What should we return if we don't return true?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
Should we return false?

bryzaguy replied:
@grimxreaper Definitely! It's super important to try and return the same data type from a function in all cases so that you don't have to go into the function to know what it will return. Granted, undefined would evaluate to false. But returning false shows your intention.

grimxreaper replied:
Really good to know, great!

bryzaguy commented on src/cleanupEquation.js:
```diff
@@ -0,0 +1,22 @@
+  nextEquation = closeLeftOpenParens(nextEquation);
+  nextEquation = addMultiplierToParensEquations(nextEquation);
+
+  if (lastEquation.length > 0) {
+    nextEquation = repeatLastEquation(lastEquation, nextEquation);
+  }
```
Is repeating the last equation cleanup? Or perhaps it's special behavior we would want to see in the onClick?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
Ah okay, we had discussed placing`repeatLastEquation` within `cleanupEquation`, but I can change that as it's not really a "clean up" action.

bryzaguy replied:
Yeah let's pull it out.

bryzaguy replied:
@grimxreaper ^

bryzaguy commented:
@grimxreaper Great job! Looks really good. I only have one correction and a few questions.

grimxreaper commented:
@bryzaguy great! Where should I look for your questions? I didn't see any comments in the code.

bryzaguy commented:
@grimxreaper I tagged you in the comments so hopefully they show up now. Let me know if they don't

bryzaguy commented:
@grimxreaper I recommend doing any fixes to the functions you've added first, so you can amend the last commit.

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,76 @@ import React from "react";
-    if ("(".includes(button)) {
-      if (result.length > 1) {
-        //these few lines below, did not work to fix another bug
+    if (isDisplayable(button)) {
```
@grimxreaper Does isDisplayable need the whole button or is only one thing from the button used? Functions are more versatile when they only ask for what they need. A side benefit is that it's easier to see from the code where the function is called that it doesn't need the whole button. Makes future refactoring easier.

grimxreaper replied:
That's really good to know, thank you for the extra information! I tried passing it only 'button.key' but it fails nearly all the tests and throws `TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))]` so I'm not sure why it seems to be reading button as undefined. I had run into this before so I just passed it `button`.

bryzaguy replied:
@grimxreaper Did you update the function when you tried passing in just the key? If you change what you pass in, you'll need to make sure the function is updated too.

grimxreaper replied:
@bryzaguy  Yes I tried that, it still throws the same error. The thing is inside of `isDisplayable` I also use `button.type` not just `button.key`.

bryzaguy replied:
@grimxreaper Hopefully this problem is addressed in my other suggestion.

grimxreaper commented:
@bryzaguy unfortunately I don't see them anywhere...

bryzaguy commented:
@grimxreaper Okay I submitted my review. Hopefully you can see them now?

grimxreaper commented on src/App.js:
```diff
@@ -2,355 +2,143 @@ import React from "react";
-        return result + button;
+      var newTempResult = result;
+      const lastCharIsOperator = "/*+-".includes(result[result.length - 1]);
+      if (lastCharIsOperator) {
```
I don't see this variable `lastCharIsOperator` anywhere in App.js. I think this is old code...

bryzaguy replied:
You're now seeing my really old comments. This one is no longer relevant.

grimxreaper commented on src/App.js:
```diff
@@ -2,355 +2,143 @@ import React from "react";
+          } catch (error) {}
+        }
+      } else {
+        const reg = /\(([0123456789/*-+.]*)\)/g;
```
I also don't see this code in App.js.

grimxreaper commented on src/App.js:
```diff
@@ -2,355 +2,143 @@ import React from "react";
-
-          } catch (error) {
-            //throw error
+            tempResultString = newTempResult + lastOperator + originalLastNum;
```
I don't see `tempResultString` anywhere in App.js. Not sure what is happening but I think this code is outdated. Sorry this is confusing!

grimxreaper commented on src/App.js:
```diff
@@ -2,355 +2,143 @@ import React from "react";
+            let expression = parenthesisToCalculate[i];
+            try {
+              let tempResult = round(evaluate(expression), 13);
+              tempResultString = tempResultString.replace(
```
`isEmptyLastOper` doesn't exist anymore, I think this is the last of the outdated code.

bryzaguy commented:
@grimxreaper I've answered your questions in the comments. Please check out the diff of your last commit and remove any unnecessary changes. Then combine that commit with the one before, since this new commit just alters what happened in the commit before.

grimxreaper committed:
extract isDisplayable, cleanupEquation, runEquation, repeatLastEquation

grimxreaper commented:
> @grimxreaper I've answered your questions in the comments. Please check out the diff of your last commit and remove any unnecessary changes. Then combine that commit with the one before, since this new commit just alters what happened in the commit before.

Done ✔️

bryzaguy commented on src/pressedAnotherOperator.js:
```diff
@@ -0,0 +1,19 @@
+const pressedAnotherOperator = (button, hasLastChar) => {
```
@grimxreaper Could be confusing to call this button since in this case button is a key and button is an object where you call the function. Might rename button to key in this case.

grimxreaper replied:
@bryzaguy So where `pressedAnotherOperator` is called it is `button.key` that is passed into it.
But inside `pressedAnotherOperator`  ` we get the syntax error `Binding member expression` when trying to replace button with button.key. Unless you mean just changing `button` to `key` in which case yes that would work. Is that what you mean?

bryzaguy replied:
@grimxreaper If you thought I meant to replace `button` with `button.key`, the error you describe makes sense. Passing in `button.key` means you don't have access to the button object anymore. I meant instead of naming the argument `button` you name it `key` since you are not passing in the button anymore, you're just passing in the key.

bryzaguy commented:
@grimxreaper Op, you squashed one commit further than I was thinking. My bad for vague instructions. You'll need to update your commit message since there's a few other functions extracted in that commit

bryzaguy commented on src/isDisplayable.js:
```diff
@@ -0,0 +1,11 @@
+const isDisplayable = (button) => {
+  const numOrDecimal = button.type === "key";
```
@grimxreaper Is there another way we can check if this is a number or decimal point? If you could figure it out using `button.key` instead, you could probably simplify what this function needs.

grimxreaper replied:
Ah okay, I tried to pass in button.key into it and changed button.type to what is below instead, but it has trouble parsing it it seems because it still throws the same error of `Binding member expression`.

`  const numOrDecimal = button.key === "." || button.key === "0" || button.key === "1" || button.key === "2" || button.key === "3" || button.key === "4" || button.key === "5" || button.key === "6" || button.key === "7" || button.key === "8" || button.key === "9";`

bryzaguy replied:
@grimxreaper Right. In order to change what you pass to the function, you have to update both the function internals, and the calling site at the same time. You can't pass in `button.key` without updating the function internals to expect only `key` coming in. I would suggest not changing the function argument until you no longer depend on `button.type` inside the function.

grimxreaper replied:
That's why I updated the function internals to button.key, and then expected it to work once I passed in button.key. /: I must be missing something about your explanation. I expected it to work since it no longer relies on button.type, if I am only using button.key inside the function (and thus passing in button.key).

bryzaguy replied:
@grimxreaper You can't pass in `button.key` until the inside of the function doesn't say `button.key` but instead says something like `key`. This is really two steps. 1) Remove any usage of button inside the function which isn't `button.key`. 2) Pass in `button.key` instead of `button` which requires updating the function internals. Don't try to combine these steps.

bryzaguy replied:
@grimxreaper Does it makes sense that you lose your reference to `button` when you pass in `button.key` and therefor get an error?

bryzaguy replied:
@grimxreaper Lastly, any way you can simplify that code for `numOrDecimal`? Maybe using `includes` or regex?

bryzaguy replied:
Can you spot there error in this code? Is it easier to see?
```is
const button = { key: "." }

isDisplayable(button.key)

function isDisplayable(button) {
   console.log(button.key)
   // .... assume the rest of the code is here.
}
```

bryzaguy replied:
Have you tried using `console.log` or `debugger`? Hoping you try to debug some of these errors.

bryzaguy replied:
@grimxreaper I think I see another issue with the `numOrDecimal` pseudo code after looking at `keypad.js`

grimxreaper replied:
I think the change I made with the latest commit is a lot cleaner than what I had and reads well to me, let me know if that isn't satisfactory, or not what you had in mind (I think it's different). Also, would a better commit message be "simplify isDisplayable, pressedAnotherOperator arguments"? Thank you! 🙌🏼

bryzaguy replied:
@grimxreaper that is a clean commit. It's exactly what I was looking for. The commit message you suggested in the last comment is great! Why didn't that end up being the message? Can we got rebase to fix it?

grimxreaper committed:
simplify isDisplayable, pressedAnotherOperator arguments

grimxreaper committed:
fix result to button and empty string when starting a new Equation, add test

grimxreaper committed:
update lastEquation  when starting a new equation

bryzaguy commented:
@grimxreaper Last commit was awesome! Now let's see if you can combine that new equation setState with the isDisplayable one.

grimxreaper committed:
combine setState within isDisplayable

grimxreaper commented:
> @grimxreaper Last commit was awesome! Now let's see if you can combine that new equation setState with the isDisplayable one.

Okay please lmk if that solution is satisfactory!

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,76 @@ import React from "react";
-    // )( -> )*(
-    const pAndp = /(\))(\()/g;
-
-    //return the value
-    return expression
-      .replace(digitAndP, "$1*$2")
-      .replace(pAndDigit, "$1*$2")
-      .replace(pAndp, "$1*$2");
-
-
-  };
-
-
-
-  calculate = () => {
-    const { result, operate, originalLastNum } = this.state;
-    let finalResult = 0;
-    var tempResult = result;
+      if (lastEquation !== "" && button.key !== "=") {
+        //start new equation
+        tempResult = "";
+      }
```
@grimxreaper Can we extract everything inside this isDisplayable if block, excluding the setState, to a function? What would be a good name?

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,76 @@ import React from "react";
-      var parenthesisToCalculate = tempResultString.match(reg) || [];
-      while (parenthesisToCalculate.length > 0) {
-        for (var i = 0; i < parenthesisToCalculate.length; i++) {
-          //we extract the first expression ex:  (2+5)
-          let expression = parenthesisToCalculate[i];
-          //We calculate the value ex: 7
 
-          try {
-            let tempResult = round(evaluate(expression), 13);
+      let finalResult = 0;
+      [nextEquation, finalResult] = runEquation(nextEquation);
```
@grimxreaper Can we extract everything in this equals pressed if block, excluding the setState, to a function? What would be a good name for it?

bryzaguy reviewed (changes requested):
@grimxreaper Once you've extracted the two functions I suggested, see if you can refactor the code in the onClick to only call setState once.

bryzaguy commented:
@grimxreaper Please let me know if you don't see the comments in code for suggested extracted functions.

bryzaguy commented:
@grimxreaper Did you mean to include App.css? With a bigger refactor it's good to stick to minimizing the changes you include. People will add formatting changes as it's own PR or when they have to touch the file or with a small, easy to review PR.

bryzaguy commented on src/App.css:
```diff
@@ -1,4 +1,4 @@
-@import url('https://fonts.googleapis.com/css2?family=Balsamiq+Sans:wght@700&display=swap');
```
Can we remove fixing formatting for App.css from this PR?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
You mean remove the formatting? I can copy paste the previous CSS version, I think that should work because I'm not sure where it was formatted to remove it manually.

bryzaguy replied:
@grimxreaper You need reset or unstash it from the commit. Make sure you don't have the file open. Your editor automatically formats on save which is why pasting the original text and s  saving won't fix it.

grimxreaper replied:
@bryzaguy  I did a git reset, and unstaged the CSS file.

grimxreaper commented:
@bryzaguy I had briefly experimented with a color change and no matter how much I "undid" it still counted the css file as changed, I tried copy pasting the previous version too and it still was counting it as changed so I wasn't sure what to do at that point.

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,54 @@ import React from "react";
       this.setState({
-        done: true,
-        result: finalResult + "",
+        result: finalValues[1] + "",
```
@grimxreaper This works!  But is there a way to have these values as named variables instead of array values?

grimxreaper replied:
How about returning in an object with labels?

bryzaguy replied:
That could work too! I think there's a nice es6 feature for turning array values into variables that you could use though...

grimxreaper replied:
I was also thinking about destructuring, okay I will do that!

grimxreaper replied:
Done

bryzaguy replied:
@grimxreaper I'm still seeing `finalresult[1]`

grimxreaper replied:
I overlooked that one, fixed now, thank you!

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,54 @@ import React from "react";
+    if (isDisplayable(button.key)) {
+      var values = updateTempResult(
+        result,
+        button.key,
```
@grimxreaper Nice job only sending .key on button!

bryzaguy commented on src/finalizeEquation.js:
```diff
@@ -0,0 +1,17 @@
+import repeatLastEquation from "./repeatLastEquation";
+import runEquation from "./runEquation";
+
+const finalizeEquation = (result, lastEquation) => {
```
@grimxreaper great job extracting this! I wonder about the name though? Finalize, while feeling like it covers the scope of the code, doesn't feel like it conveys the most important job of this function. What might be a better name?

grimxreaper replied:
It changes the equation, but I wasn't sure that 'alterEquation' would be a good name. I'm guessing the most important job of this function is that it returns the correct equation...

bryzaguy replied:
@grimxreaper Think of this function as its own npm package. What would you call it? When would you tell someone to use it? What problem would they have that this solves?

grimxreaper replied:
I would call it updateEquation or editEquation...?
I would tell someone to use it when they want to ensure that the equation in the display has been polished and edited to reflect the correct equation. Whether that edit requires the equation to be repeated, ran, or cleaned up.

bryzaguy replied:
@grimxreaper If I gave someone that function I would tell them it runs the equation! Cleaning it up is just so you can run it.

grimxreaper replied:
Ah okay yes! I was hesitating to call it that because we already have a runEquation function which also definitely runs the equation. Do you suggest renaming that one?

grimxreaper replied:
I renamed it, I think it makes more sense now!

bryzaguy commented on src/updateTempResult.js:
```diff
@@ -0,0 +1,26 @@
+import pressedAnotherOperator from "./pressedAnotherOperator";
+
+const updateTempResult = (result, key, hasLastChar, lastEquation) => {
```
@grimxreaper Good job extracting this! Not sure about the name though. The word result doesn't conjure an image of the place that's updated since it can either have a result or just show the equation in process. Definitely don't dig "Temp" in the name since we likely only named it that because result was already taken. What would be a better name?

Also, is there a way we can simplify the parameter list / return value?

grimxreaper replied:
How about 'updateDisplay'? 
Not sure how I would simplify the parameter list or what is returned...I'll keep thinking.

bryzaguy replied:
@grimxreaper Great name suggestion! Looking forward to the simplification changes.

grimxreaper replied:
@bryzaguy Is it okay if I make a new commit with the name changes? Or do you want me to go back in history and edit?

bryzaguy replied:
Feel free to do a new commit. If this were a PR for an open source project we'd want to make sure our commits condensed. @grimxreaper

bryzaguy reviewed (changes requested):
@grimxreaper Commented where I'm hoping you make some changes. Let me know if you don't see them. Also, your commit message for extracting updateTempResult says "simplify isDisplayable". If the code you simplified is in the extracted function you might drop the "isDisplayable" so it doesn't seem like you updated two different areas. (i.e. "extract [function-name], simplify")

bryzaguy commented:
> @bryzaguy I had briefly experimented with a color change and no matter how much I "undid" it still counted the css file as changed, I tried copy pasting the previous version too and it still was counting it as changed so I wasn't sure what to do at that point.

You can leave the App.css unstaged so it doesn't get committed.

grimxreaper commented:
@bryzaguy I used git rebase to edit the commit several times with no luck. Even though I quit, it now shows "Your branch and 'origin/refactor-migrate-changeKeys' have diverged, and have 2 and 2 different commits each". Gah, I'm not sure what to do now, I could git pull but I don't want to screw up the commits.

bryzaguy commented:
You're so close!
`git push --force`
@grimxreaper ^

grimxreaper committed:
extract updateTempResult and finalizeEquation

grimxreaper committed:
update updateDisplay and runEquation to descriptive function names

bryzaguy commented on src/updateDisplay.js:
```diff
@@ -0,0 +1,26 @@
+import pressedAnotherOperator from "./pressedAnotherOperator";
+
+const updateDisplay = (result, key, hasLastChar, lastEquation) => {
+  var tempResult = result;
```
Is it necessary to create tempResult or currentBtn variables?

grimxreaper replied:
I replaced currentBtn to key and tempResult to result within the `updateDisplay` function and everything works fine. I thought they needed to be separated but it doesn't look like it. Although I cannot use 'result' in App.js to replace tempResult in the destructuring because result has already been declared. Perhaps I can use currentResult? What do you think?

bryzaguy replied:
I'm only concerned about the function internals. App.JS can stay the same

grimxreaper replied:
Ah okay, got it.

bryzaguy commented on src/updateDisplay.js:
```diff
@@ -0,0 +1,26 @@
+import pressedAnotherOperator from "./pressedAnotherOperator";
+
+const updateDisplay = (result, key, hasLastChar, lastEquation) => {
```
@grimxreaper When extracting a function, look really carefully to see what code you can move into the function so you don't have to pass the result of that code in or return it. Also, determine if it makes sense that the function would be responsible for the code you move into it. (e.g. if I didn't know what was inside the function, would I be surprised if this code ran inside) There's not always a right answer, but it is a valuable skill you can improve.

bryzaguy replied:
Any guesses what you might pull into the function?

grimxreaper replied:
Ah okay, yes. `hasLastChar`?

bryzaguy replied:
That's a good catch! There's more though

grimxreaper replied:
- `hasLastChar` is also used by `pressedAnotherOperator`, I moved it into the updateDisplay function, and the tests and app seem to work fine, but why? How does `pressedAnotherOperator` have access to `hasLastChar`?
- I moved this line (`const { result, lastEquation } = this.state;`) into `updateDisplay` so we wouldn't need to pass result & lastEquation in, but I would need to call them something else when they are used within the `updateDisplay` function because 'result' is a constant.
- I would also need to keep `const { result, lastEquation } = this.state;` within the onClick function in App.js because they are used elsewhere. So I'm not sure that that is worthwhile, is this what you were thinking?

bryzaguy replied:
Looks like you import and use `pressedAnotherOperator` in the `updateDisplay` file, so as long as `hasLastChar` can be found there, you should be good. The "one more" I was referring to wasn't for simplifying any parameters beyond `hasLastChar`. Any guesses at what other code could be moved into `updateDisplay`?

grimxreaper replied:
Well we can't move the setState in there right, so hmm, are you hinting at the next block over? The `(button.key === "=")` if block?

bryzaguy replied:
@grimxreaper I'm hinting at the return value

grimxreaper replied:
You're saying the return value should be moved into `updateDisplay`...I'm confused 😅 do you mean this `var [tempResult, currentBtn]` inside of the onClick? I'm not seeing what you mean...

bryzaguy replied:
Look carefully at what you return and how it's used. Is there a way you can return a single value rather than two values?

bryzaguy replied:
@grimxreaper

grimxreaper replied:
How about this, I combined `tempResult` and `key` by returning this from `updateDisplay`:  ` return result + key`)
and changing this in the `onClick`:

```
var currResult = updateDisplay(
          result,
          button.key,
          lastEquation
        );

        this.setState({
          result: currResult,
          lastEquation: "",
        });
```

bryzaguy replied:
Yes!! Although is it really the currResult (don't be afraid to spell out current) or is it the nextResult since you just updated it?

grimxreaper replied:
Oh true that! I just named it real fast and didn't think about that. Also I feel silly for not seeing this earlier. Will update now.

bryzaguy replied:
All good. We are training your eyes to catch these things.

grimxreaper committed:
destructure runEquation return variables

bryzaguy commented on src/Components/keypads.js:
```diff
@@ -0,0 +1,25 @@
+const keypads = [
+    { type: "operation", key: "(" },
```
Do we even use the type property?

grimxreaper replied:
We don't anymore, I will remove it.

bryzaguy commented on src/isDisplayable.js:
```diff
@@ -0,0 +1,11 @@
+const isDisplayable = (key) => {
```
@grimxreaper Would it be possible to add data to keypads rather than use this function?

grimxreaper replied:
Hm, I'm not sure what you mean...which data? Do you mean replacing lines 2-5 in `isDisplayable`?

bryzaguy replied:
I mean, could we store some data on keypads so that we no longer use this function at all? This function evaluates  whether a key is displayable but we know  when we look at keypads we know which ones are and could put something there that we check instead of calling a function. Hopefully I'm making sense? @grimxreaper

grimxreaper replied:
Yes that makes sense, I'm thinking of including something like this in keypads but not sure how because keypads is an array of objects, so I can't insert this there, I'd have to make another function I think within that `keypads` component.

```
  if (key) {
    return true;
  }
  return false;
```

grimxreaper replied:
But I don't think that's what you're getting at. I think you probably mean adding a property onto keypads for every key except '=, AC, CE'? Something like this:

<img width="276" alt="Screen Shot 2021-07-09 at 9 53 38 AM" src="https://user-images.githubusercontent.com/64915527/125088626-9b00b680-e09b-11eb-8813-f23204505221.png">

And in App.js this is how I referred to the keypads without using the `isDisplayable` function.

<img width="414" alt="Screen Shot 2021-07-09 at 10 08 17 AM" src="https://user-images.githubusercontent.com/64915527/125090699-a5bc4b00-e09d-11eb-87a8-25ee53ac7d6a.png">

bryzaguy replied:
@grimxreaper Yes!! Can we think of a better name for that `operation` property?

grimxreaper replied:
@bryzaguy hmmm, how about, expression? or equation? We aren't including the '=' so I'm not sure if calling it any of these names is clear enough.

bryzaguy replied:
@grimxreaper What is the purpose of the new property? What does it mean if a key has that property set to true?

grimxreaper replied:
The purpose of it is to indicate that it should be displayed as part of the equation...

bryzaguy replied:
How do you want your code to read? It will show up as `if (button.something) { ...` where "something" is a name that lets you know this button should be displayed as part of equation. Could be similar to what it's replacing.

bryzaguy replied:
@grimxreaper

grimxreaper replied:
Great question. `if (button.display)` `if (button.show)` does either read well?

bryzaguy replied:
Those are much better. Try reading it... "if button display" and "if button show" doesn't tell me that they are booleans. They could be data and we are checking that the data is there. A good rule of thumb for Boolean properties is to make them adjectives, sometimes prefixed with a verb like "is". Do you think `if (button.isShowable) { ...` or `if (button.isDisplayable) { ...` would be easier to understand?

grimxreaper replied:
If they are Boolean properties, make them adjectives prefixed with an "is", okay great! Yes I think I'll stick with the second one.

bryzaguy replied:
Perfect

bryzaguy replied:
I should correct myself. If the Boolean represents a characteristic or quality of something, make it an adjective. If you're saying what it literally is, nouns are cool (i.e. "isBoat"). Using a prefix verb like "is", "was", "has" is good for both cases so if conditions read well.

bryzaguy reviewed (changes requested):
Great function name changes! Made a few more suggestions. Let me know if you don't see them.

bryzaguy commented:
You're getting so close! Excited to see your next changes!

bryzaguy commented:
One minor note. Your two last commit messages would be more helpful if they contained the function names they're referring to.

grimxreaper committed:
add isDisplayable to keypad, simplify updateDisplay return value

bryzaguy commented:
I should correct myself. If the Boolean represents a characteristic or quality of something, make it an adjective. If you're saying what it literally is, nouns are cool (i.e. "isBoat"). Using a prefix verb like "is", "was", "has" is good for both cases so if conditions read well.

bryzaguy commented:
As for your commit message. Minor nit: "returned value" is generally referred to as "return value". Could read as "Add isDisplayable to keypad, simplify updateDisplay return value". Do you notice the extra info in there? It answers questions like when was "isDisplayable" prop introduced?".

grimxreaper commented:
Oooh yes mentioning the property introduced, or adding the function names helps clarify when things were added!

bryzaguy commented on src/App.js:
```diff
@@ -2,355 +2,44 @@ import React from "react";
-        for (var i = 0; i < parenthesisToCalculate.length; i++) {
-          //we extract the first expression ex:  (2+5)
-          let expression = parenthesisToCalculate[i];
-          //We calculate the value ex: 7
-
-          try {
-            let tempResult = round(evaluate(expression), 13);
-
-            //We need to replace the expression by the calculation
-            tempResultString = tempResultString.replace(expression, tempResult);
-
-          } catch (error) {
-            //throw error
-          }
-        }
-
-        parenthesisToCalculate = tempResultString.match(reg) || [];
-      }
-      //In tempResultString we have the last expression withtout any ()
-
-      try {
-       finalResult = round(evaluate(tempResultString), 13);
-      } catch (error) {
-        //throw error
-      }
       this.setState({
-        done: true,
         result: finalResult + "",
+        lastEquation: nextEquation,
+      });
+    } else if (button.key === "AC") {
+      this.setState({
+        result: "",
+        lastEquation: "",
+      });
+    } else if (button.key === "CE") {
+      this.setState({
+        result: result.slice(0, -1),
       });
     }
```
@grimxreaper Can we refactor this so there is a single setState at the end?

grimxreaper replied:
@bryzaguy  It isn't super clear from all the code referenced, do you mean refactoring to one single setState in the onClick?

bryzaguy replied:
You got it! Yeah I was essentially highlighting the internals of onClick.

grimxreaper replied:
So I feel like this should work but I get back loads of red text that say "newTempResult.match is not a function" but it doesn't give me the actual error. Could you give me a hint as to what I would need to fix? 

<img width="602" alt="Screen Shot 2021-07-11 at 1 27 14 PM" src="https://user-images.githubusercontent.com/64915527/125204578-fe314b00-e24b-11eb-8d8c-884e1f735f29.png">

<img width="543" alt="Screen Shot 2021-07-11 at 1 27 24 PM" src="https://user-images.githubusercontent.com/64915527/125204587-01c4d200-e24c-11eb-81c4-1fcdd7d02389.png">

bryzaguy replied:
I want to see a commit where you successfully reduce it to a single setState. You are making it harder by extracting a function at the same time. This is why small steps are important. Don't skip steps!

Do you know where the error is occurring? Have you tried using console logs, debugger, or commenting out code?

bryzaguy replied:
Imagine a bug occurred that you didn't catch, then someone has to figure out what about your commit broke things. If it's in separate commits, it's easier to identify with git bisect.

grimxreaper replied:
How would I reduce it to a single setState without extracting a function? It seems that it is extracting the function that would make a single setState possible...

bryzaguy replied:
Nope. Totally possible without extracting. In fact, necessary in order to be able to extract it.

bryzaguy replied:
The idea is you want a function that has everything currently in the onClick except for the react state stuff. In order to extract it, you must first "move" the setState's out of the way. Only then can you extract the code safely. In fact, by doing so you may avoid the bug you're seeing now.

grimxreaper replied:
Do you mean something like this?
<img width="569" alt="Screen Shot 2021-07-11 at 2 30 49 PM" src="https://user-images.githubusercontent.com/64915527/125206319-a0edc780-e254-11eb-9dd8-1e8fad964820.png">

bryzaguy replied:
That's better. Any guesses why you might have issues with that?

bryzaguy replied:
Do you notice the squiggly line?

bryzaguy replied:
What if you tried only removing one setState and seeing if things still work? Then removing another. Instead of removing them all and seeing what happens.

grimxreaper replied:
> Do you notice the squiggly line?

Yes, but I'm not sure what to do about that since the variable name will need to be the same for me to only setState once, or that's my understanding at least.

grimxreaper replied:
> What if you tried only removing one setState and seeing if things still work? Then removing another. Instead of removing them all and seeing what happens.

That's a good idea, I tried it, and it doesn't actually seem to be updating state, even when I go one at a time.

bryzaguy reviewed (changes requested):
Great job! We are soooo close to the end. Look how small our onClick is getting! A few changes left.

bryzaguy commented:
@grimxreaper The commit messages are better! Really try to front load the specifics (function names in this case) to avoid truncation.

grimxreaper commented:
> @grimxreaper The commit messages are better! Really try to front load the specifics (function names in this case) to avoid truncation.

How would you rewrite this to front load the function name?
"destructuring return variables in runEquation function"

Like "runEquation return variables destructuring"?

bryzaguy commented:
Also "Applying this commit will...destructuring?"

bryzaguy commented:
> > @grimxreaper The commit messages are better! Really try to front load the specifics (function names in this case) to avoid truncation.
> 
> 
> 
> How would you rewrite this to front load the function name?
> 
> "destructuring return variables in runEquation function"
> 
> 
> 
> Like "runEquation return variables destructuring"?

I replied in text because I couldn't paste a screenshot. Just move things around a little to “add `comma separated func names` for descriptive function names” and “destructure runEquation return variables”

grimxreaper committed:
reduce onClick setState calls, minor changes

grimxreaper committed:
extract handleButtonPress
